File: 8_kyu/Total_amount_of_points.py
from collections.abc import Iterable

# Converts match result into points.
# Argument
#   String in format "home:other" with the scores.
# Return
#   Points 0, 1 or 3 reached by the home team.
def match_points(result: str) -> int:
    home, other = result.split(":",3)
    delta = int(home)-int(other)
    if delta == 0:
        return 1 # tie
    elif delta < 0:
        return 0 # loss
    else:
        return 3 # win

# Calculates the point sum of all matches.
# Argument
#   Sequence of match results "home:other"
# Return
#   Sum of all points reached in the matches
def points(games: Iterable[str]) -> int:
	return sum(match_points(match_result) for match_result in games)

File: 8_kyu/Total_amount_of_points.kt
// Returns the total points of all games.
fun points(games: List<String>) = games.sumOf { points(parseGame(it)) }

// Pair of the two scores in a game result.
private typealias gameResult = Pair<Int, Int>

// Parses game result from string "score:score" into gameResult integer pair.
private fun parseGame(game: String): gameResult {
    val splits = game.split(":", limit = 3)
    if (splits.size != 2) throw IllegalArgumentException()
    return splits[0].trim().toInt() to splits[1].trim().toInt()
}

// Returns the points of the first team from a game result.
private fun points(result: gameResult): Int {
    val delta = result.first - result.second
    return when {
        delta > 0 -> 3
        delta < 0 -> 0
        else -> 1
    }
}
